### ISSUE TITLE
Completion with Irony issue: Invalid function: (cb callback)

### ISSUE THREAD

**Issue by ANDRON94** [Sarcasm/irony-mode#426]
Hi,
After the automatic update of Melpa packages company completion with irony stopped working.
I've tried to complete "std::str[ing]" and got next error messages:

> Company: An error occurred in auto-begin
> Invalid function: (cb callback)
> Cannot complete at point

Also, attach stack trace here:

> Debugger entered--Lisp error: (invalid-function (cb callback))
>   (cb callback)()
>   irony-completion-candidates-async(#[257 "\300\301!!\207" [#[257 "\300\305\303\300\242\"\240\210\302\304!\240\210\301 \207" [(((:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["str" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"]))) #[0 "\303\242?\205
> \301\300\302\242\237!!\207" [#[257 "\300\301\"\207" [apply append] 4 "\n\n(fn VALUES)"] #[257 "\211\203\301\242\307=\204\301\240\207\306\f\300\310!BC
> \311\312DC\216\313\302\303\304\305$)\207" ["str" ... #<buffer test.cpp> #<window 248 on test.cpp> 418 21543 ... done company--preprocess-candidates funcall #[0 "?\205\n\303\211\211\207" [company-candidates company-backend company-candidates-cache nil] 3 "\n\n(fn)"] company-idle-begin company-backend company-candidates-cache] 6 "\n\n(fn CANDIDATES)"] ((nil)) ((...))] 3 "\n\n(fn)"] (nil) (:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["str" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"]) #[257 "\301\203\f\300\304!\262)\302\203!\305\306\307\310\311\312\300!\313\"\314\315%\"\262\207" [company-irony nil t company-backend company--preprocess-candidates mapcar make-byte-code 257 "\301\302\300#\207" vconcat vector [propertize company-backend] 5 "\n\n(fn STR)"] 8 "\n\n(fn CANDIDATES)"] delq] 5 "\n\n(fn RES)"] company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)"] "str" case-insensitive)
>   #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["str" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"](#[257 "\300\305\303\300\242\"\240\210\302\304!\240\210\301 \207" [(((:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["str" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"]))) #[0 "\303\242?\205
> \301\300\302\242\237!!\207" [#[257 "\300\301\"\207" [apply append] 4 "\n\n(fn VALUES)"] #[257 "\211\203\301\242\307=\204\301\240\207\306\f\300\310!BC
> \311\312DC\216\313\302\303\304\305$)\207" ["str" (nil) #<buffer test.cpp> #<window 248 on test.cpp> 418 21543 (company-irony-c-headers company-irony) done company--preprocess-candidates funcall #[0 "?\205\n\303\211\211\207" [company-candidates company-backend company-candidates-cache nil] 3 "\n\n(fn)"] company-idle-begin company-backend company-candidates-cache] 6 "\n\n(fn CANDIDATES)"] ((nil)) (((:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["str" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"])))] 3 "\n\n(fn)"] (nil) (:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["str" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"]) #[257 "\301\203\f\300\304!\262)\302\203!\305\306\307\310\311\312\300!\313\"\314\315%\"\262\207" [company-irony nil t company-backend company--preprocess-candidates mapcar make-byte-code 257 "\301\302\300#\207" vconcat vector [propertize company-backend] 5 "\n\n(fn STR)"] 8 "\n\n(fn CANDIDATES)"] delq] 5 "\n\n(fn RES)"])
>   #[257 "\302C\303\304\300\"C\305\306\307\310\311\301$\312\"\313\314%\300\211\205a\211@\302\242B\240\210\242@A\305\315\316\310\311			%\317\"\320\321%\242\322=\204Q\211!\210\202XA\211!\266\266A\266\202\202\262\207" [(((:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["str" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"]) . #[257 "\301\203\f\300\304!\262)\302\203!\305\306\307\310\311\312\300!\313\"\314\315%\"\262\207" [company-irony nil t company-backend company--preprocess-candidates mapcar make-byte-code 257 "\301\302\300#\207" vconcat vector [propertize company-backend] 5 "\n\n(fn STR)"] 8 "\n\n(fn CANDIDATES)"])) #[257 "\300\301\"\207" [apply append] 4 "\n\n(fn VALUES)"] nil mapcar car make-byte-code 0 "\303\242?\205
> \301\300\302\242\237!!\207" vconcat vector [] 3 "\n\n(fn)" 257 "\300\305\303\300\242\"\240\210\302\304!\240\210\301 \207" [delq] 5 "\n\n(fn RES)" :async] 19 "\n\n(fn CALLBACK)"](#[257 "\211\203\301\242\307=\204\301\240\207\306\f\300\310!BC
> \311\312DC\216\313\302\303\304\305$)\207" ["str" (nil) #<buffer test.cpp> #<window 248 on test.cpp> 418 21543 (company-irony-c-headers company-irony) done company--preprocess-candidates funcall #[0 "?\205\n\303\211\211\207" [company-candidates company-backend company-candidates-cache nil] 3 "\n\n(fn)"] company-idle-begin company-backend company-candidates-cache] 6 "\n\n(fn CANDIDATES)"])
>   company--fetch-candidates("str")
>   company-calculate-candidates("str")
>   #[0 "\n\203	\nC\202\n\211\205\212\211@\3009\204\306!\203+\307!\205/\310\311!)\202/\312\311\"\240\210\300\242\203\203\313\300\242!\203}\314\300\242!\211\301\315\f!\240\210\301\242:\204f
> \203X\316\317!\210\301\242\320=\203}\321\322!\210\202}
> \203n\300\242\323\301\242!\210\324\325\326 \"\210\327\330!\210\331\332\301\242\"\210A\266\202\202\n\207" [(("str" . t)) (nil) company-backend company-backends company-prefix company--manual-action functionp company--maybe-init-backend company-call-backend prefix company--multi-backend-adapter company--good-prefix-p company--prefix-str company-calculate-candidates message "No completion found" t company-cancel unique company-update-candidates run-hook-with-args company-completion-started-hook company-explicit-action-p company-call-frontends show throw --cl-block-nil-- company--manual-prefix] 6 "\n\n(fn)"]()
>   funcall(#[0 "\n\203	\nC\202\n\211\205\212\211@\3009\204\306!\203+\307!\205/\310\311!)\202/\312\311\"\240\210\300\242\203\203\313\300\242!\203}\314\300\242!\211\301\315\f!\240\210\301\242:\204f
> \203X\316\317!\210\301\242\320=\203}\321\322!\210\202}
> \203n\300\242\323\301\242!\210\324\325\326 \"\210\327\330!\210\331\332\301\242\"\210A\266\202\202\n\207" [(("str" . t)) (nil) company-backend company-backends company-prefix company--manual-action functionp company--maybe-init-backend company-call-backend prefix company--multi-backend-adapter company--good-prefix-p company--prefix-str company-calculate-candidates message "No completion found" t company-cancel unique company-update-candidates run-hook-with-args company-completion-started-hook company-explicit-action-p company-call-frontends show throw --cl-block-nil-- company--manual-prefix] 6 "\n\n(fn)"])
>   company--begin-new()
>   company--perform()
>   #[0 "\301 \210\207" [company-candidates company--perform] 1 "\n\n(fn)"]()
>   funcall(#[0 "\301 \210\207" [company-candidates company--perform] 1 "\n\n(fn)"])
>   company-auto-begin()
>   company-idle-begin(#<buffer test.cpp> #<window 248 on test.cpp> 418 21543)
>   apply(company-idle-begin (#<buffer test.cpp> #<window 248 on test.cpp> 418 21543))
>   byte-code("r\301\302H\303H\"\210)\301\207" [timer apply 5 6] 4)
>   timer-event-handler([t 22962 50075 502235 nil company-idle-begin (#<buffer test.cpp> #<window 248 on test.cpp> 418 21543) nil 929000])

**Comment by Sarcasm** [comment]
Are you sure both irony-mode and company-irony are up-to-date?
If only one of them has been updated, you can have this sort of errors because a function signature changed.

**Comment by ANDRON94** [comment]
Ok, I'll check it. Thanks.

**Comment by ANDRON94** [comment]
Hi, I update emacs packages to the latest versions:

> company-20170828.414
> company-irony-20170905.1346
> irony-20170905.1316

reinstall irony server but still have the same issue.
Here is backtrace:

> Debugger entered--Lisp error: (invalid-function (cb callback))
>   (cb callback)()
>   irony-completion-candidates-async(#[257 "\300\301!!\207" [#[257 "\300\305\303\300\242\"\240\210\302\304!\240\210\301 \207" [(((:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["stri" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"]))) #[0 "\303\242?\205
> \301\300\302\242\237!!\207" [#[257 "\300\301\"\207" [apply append] 4 "\n\n(fn VALUES)"] #[257 "\211\203\301\242\307=\204\301\240\207\306\f\300\310!BC
> \311\312DC\216\313\302\303\304\305$)\207" ["stri" ... #<buffer test.cpp> #<window 3 on test.cpp> 426 21544 ... done company--preprocess-candidates funcall #[0 "?\205\n\303\211\211\207" [company-candidates company-backend company-candidates-cache nil] 3 "\n\n(fn)"] company-idle-begin company-backend company-candidates-cache] 6 "\n\n(fn CANDIDATES)"] ((nil)) ((...))] 3 "\n\n(fn)"] (nil) (:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["stri" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"]) #[257 "\301\203\f\300\304!\262)\302\203!\305\306\307\310\311\312\300!\313\"\314\315%\"\262\207" [company-irony nil t company-backend company--preprocess-candidates mapcar make-byte-code 257 "\301\302\300#\207" vconcat vector [propertize company-backend] 5 "\n\n(fn STR)"] 8 "\n\n(fn CANDIDATES)"] delq] 5 "\n\n(fn RES)"] company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)"] "stri" case-insensitive)
>   #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["stri" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"](#[257 "\300\305\303\300\242\"\240\210\302\304!\240\210\301 \207" [(((:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["stri" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"]))) #[0 "\303\242?\205
> \301\300\302\242\237!!\207" [#[257 "\300\301\"\207" [apply append] 4 "\n\n(fn VALUES)"] #[257 "\211\203\301\242\307=\204\301\240\207\306\f\300\310!BC
> \311\312DC\216\313\302\303\304\305$)\207" ["stri" (nil) #<buffer test.cpp> #<window 3 on test.cpp> 426 21544 (company-irony-c-headers company-irony) done company--preprocess-candidates funcall #[0 "?\205\n\303\211\211\207" [company-candidates company-backend company-candidates-cache nil] 3 "\n\n(fn)"] company-idle-begin company-backend company-candidates-cache] 6 "\n\n(fn CANDIDATES)"] ((nil)) (((:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["stri" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"])))] 3 "\n\n(fn)"] (nil) (:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["stri" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"]) #[257 "\301\203\f\300\304!\262)\302\203!\305\306\307\310\311\312\300!\313\"\314\315%\"\262\207" [company-irony nil t company-backend company--preprocess-candidates mapcar make-byte-code 257 "\301\302\300#\207" vconcat vector [propertize company-backend] 5 "\n\n(fn STR)"] 8 "\n\n(fn CANDIDATES)"] delq] 5 "\n\n(fn RES)"])
>   #[257 "\302C\303\304\300\"C\305\306\307\310\311\301$\312\"\313\314%\300\211\205a\211@\302\242B\240\210\242@A\305\315\316\310\311			%\317\"\320\321%\242\322=\204Q\211!\210\202XA\211!\266\266A\266\202\202\262\207" [(((:async . #[257 "\301\302\303\304\305\306!\307\"\310\311%\300\312 #\207" ["stri" irony-completion-candidates-async make-byte-code 257 "\300\301!!\207" vconcat vector [company-irony--make-candidates] 4 "\n\n(fn CANDIDATES)" company-irony--get-matching-style] 8 "\n\n(fn CALLBACK)"]) . #[257 "\301\203\f\300\304!\262)\302\203!\305\306\307\310\311\312\300!\313\"\314\315%\"\262\207" [company-irony nil t company-backend company--preprocess-candidates mapcar make-byte-code 257 "\301\302\300#\207" vconcat vector [propertize company-backend] 5 "\n\n(fn STR)"] 8 "\n\n(fn CANDIDATES)"])) #[257 "\300\301\"\207" [apply append] 4 "\n\n(fn VALUES)"] nil mapcar car make-byte-code 0 "\303\242?\205
> \301\300\302\242\237!!\207" vconcat vector [] 3 "\n\n(fn)" 257 "\300\305\303\300\242\"\240\210\302\304!\240\210\301 \207" [delq] 5 "\n\n(fn RES)" :async] 19 "\n\n(fn CALLBACK)"](#[257 "\211\203\301\242\307=\204\301\240\207\306\f\300\310!BC
> \311\312DC\216\313\302\303\304\305$)\207" ["stri" (nil) #<buffer test.cpp> #<window 3 on test.cpp> 426 21544 (company-irony-c-headers company-irony) done company--preprocess-candidates funcall #[0 "?\205\n\303\211\211\207" [company-candidates company-backend company-candidates-cache nil] 3 "\n\n(fn)"] company-idle-begin company-backend company-candidates-cache] 6 "\n\n(fn CANDIDATES)"])
>   company--fetch-candidates("stri")
>   company-calculate-candidates("stri")

My test compilation database:

> Compilation Database: irony-cdb-clang-complete
> 
>   Working Directory: /home/user/Desktop/
>   Compile Options:   -std=c++14

**Comment by ANDRON94** [comment]
Hi, forget to add that flycheck-irony works as expected and shows errors correctly.

**Comment by Sarcasm** [comment]
Does it work if you change the definition of `irony-completion-candidates-async` to:

```el
(defun irony-completion-candidates-async (callback &optional prefix style)
  (irony--aif (irony-completion-symbol-bounds)
      (irony--run-task-asynchronously
       (irony--candidates-task nil (car it) prefix style)
       (lambda (candidates-result)
         (funcall callback (irony-completion--filter-candidates
                            (irony-iotask-result-get candidates-result)))))
    (funcall callback nil)))
```

I don't understand the issue because I don't reproduce it, and the diagnostic callback mechanism is the same as for completion, so I would expect neither to work, or both, but not only one. :/

**Comment by ANDRON94** [comment]
Yeap, it's helped me! Thanks.
 But what is wrong with "lexical-let"?

**Comment by Sarcasm** [comment]
Oh great. I would expect nothing to be wrong with it, it seems to work for syntax checking... However, it is no longer necessary because the file use lexical binding, so that's why I tried to remove it.

I would be curious to understand but I will probably remove the lexical-let either way.

**Comment by ANDRON94** [comment]
I call it "magic code" =)

**Comment by Sarcasm** [comment]
I'm wondering if your configuration has anything magical.... :)

**Comment by ANDRON94** [comment]
Maybe you're right  ;)

// Irony

> (defun my-irony-mode-hook ()
>   (define-key irony-mode-map [remap completion-at-point]
>     'irony-completion-at-point-async)
>   (define-key irony-mode-map [remap complete-symbol]
>     'irony-completion-at-point-async))
> 
> (use-package irony
>   :ensure t
>   :commands (irony-mode)
>   :init
>   (add-hook 'c++-mode-hook 'irony-mode)
>   (add-hook 'c-mode-hook 'irony-mode)
>   :config
>   (add-hook 'irony-mode-hook 'my-irony-mode-hook)
>   (add-hook 'irony-mode-hook 'irony-cdb-autosetup-compile-options))

// Company irony

>   (use-package company-irony
>     :ensure t
>     :config
>     (use-package company-irony-c-headers :ensure t)
>     (setq company-backends (delete 'company-semantic company-backends))
>     (add-to-list 'company-backends '(company-irony-c-headers company-irony))
>     (with-eval-after-load 'irony
>       (add-hook 'irony-mode-hook 'company-irony-setup-begin-commands)))

// Flycheck irony

>   (use-package flycheck-irony ; Integrate flycheck with Irony
>     :ensure t
>     :init (add-hook 'flycheck-mode-hook 'flycheck-irony-setup))

**Comment by Sarcasm** [comment]
Yeah, I actually spied on your Github and did not find anything shaddy.
I would be curious if it happens if you comment your use-package config, and configure irony-mode manually (as per the README instruction).

I'm thinking maybe some sort of lazy/late evaluation could cause this (it's a bit of a random thought).

**Comment by ANDRON94** [comment]
Anyway, I glad that it was fixed. I'll be waiting for your next release with (cb callback) removed and other nice features. Thank you, for your awesome irony package!

**Comment by Sarcasm** [comment]
Fix committed as 6579e7be6b571c7bd96f082669469aafd97598c2.

**Comment by ANDRON94** [comment]
Thank you!
